### PR TITLE
Implement k_yield() replica in arch/xtos-wrapper/include/kernel.h and remove wait_for_interrupt()

### PR DESCRIPTION
**Pull request by jsarha**

### CODE REVIEW

File: src/arch/xtos-wrapper/include/kernel.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
Comment (kv2019i) [inline src/arch/xtos-wrapper/include/kernel.h:1]: I'd reword the commit to explain what the commit does. Once merged, people won't see the pull-request context and "first commit" won't make sense.

Comment (kv2019i) [reply src/arch/xtos-wrapper/include/kernel.h:1]: Maybe refer to the place in source where the implementation is taken from and state this is implementing k_yield() Zephyr interface using XTOS.

File: src/arch/xtensa/schedule/task.c [hunk 2]
```diff
@@ -16,7 +16,6 @@
 #include <sof/lib/cache.h>
 #include <sof/lib/cpu.h>
 #include <sof/lib/memory.h>
-#include <sof/lib/wait.h>
```
Comment (kv2019i) [inline src/arch/xtensa/schedule/task.c:19]: commit typo: r_yield() -> k_yield()

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @jsarha ! Looks good. Some minor changes to the commit messages, but otherwise ready to go.

**Review by dbaluta — APPROVED** [review]

File: src/arch/xtensa/schedule/task.c (hunk 2 above)
Comment (jsarha) [reply src/arch/xtensa/schedule/task.c:19]: Oops. I wonder why this did not fail in CI build. Hope there are no more such errors.

**Review by kv2019i — APPROVED** [review]
Good to go!

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
@lgirdwood looks good to me. Cannot merged this because of some tests not running.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
CI showing a testbench issue - the issue looks unrelated top this PR. @singalsu any comments ?

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by singalsu** [comment]
> CI showing a testbench issue - the issue looks unrelated top this PR. @singalsu any comments ?

It looks like IIR test memory leak. I'll get this PR and try if I'm able to repeat. It doesn't happen with git main for me.

**Comment by jsarha** [comment]
When I run the test I see the same error as seen in CI, but I see it whether or not I have my changes in. I see if I can investigate it further tomorrow.

**Comment by jsarha** [comment]
I do not really know how to investigate this further, but it looks to me that "check-suspend-resume-with-audio.sh" is pretty unstable at the moment, with or without this PR. I have not been able to have a single successful run on my Up-ext-i11 in the whole morning.

The test case script always fails on this:
error: in logger_read(), fread(..., /sys/kernel/debug/sof/trace) failed: Operation not permitted(-1)

At the same time the dmesg looks like this:

Mar 25 11:50:29 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Failed to setup widget PIPELINE.10.DMIC0.IN
Mar 25 11:50:29 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed widget list set up for pcm 6 dir 1
Mar 25 11:50:29 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at snd_soc_pcm_component_hw_params on 0000:00:1f.3: -110
Mar 25 11:50:29 jhary kernel:  DMIC: ASoC: __soc_pcm_hw_params() failed (-110)
Mar 25 11:50:29 jhary kernel:  DMIC: ASoC: dpcm_fe_dai_hw_params failed (-110)
Mar 25 11:50:29 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: free stream 6 dir 1
Mar 25 11:50:29 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: close stream 6 dir 1
Mar 25 11:50:30 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx timed out for 0x90040000 (msg/reply size: 60/12)
Mar 25 11:50:30 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware exception
Mar 25 11:50:30 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: fail in sof_ipc_trace_update_filter -110
Mar 25 11:50:31 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx timed out for 0x90040000 (msg/reply size: 60/12)
Mar 25 11:50:31 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware exception
Mar 25 11:50:31 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: fail in sof_ipc_trace_update_filter -110
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx timed out for 0x90050000 (msg/reply size: 8/12)
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware exception
**Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: DMA_TRACE_FREE failed with error: -110**
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x40010000
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx timed out for 0x40010000 (msg/reply size: 76/12)
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware exception
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ctx_save IPC error: -110, proceeding with suspend
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Debug PCIR: 00000000 at  00000044
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Debug PCIW: 00000010 at  00000044
Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: fw_state change: 6 -> 0

The bold line is the one that does not appear with the other suspend resume cycles. It is coming from sound/soc/sof/trace.c:582 after a failed sof_ipc_tx_message() call. Should I create a new issue out of this?

**Comment by lgirdwood** [comment]
> When I run the test I see the same error as seen in CI, but I see it whether or not I have my changes in. I see if I can investigate it further tomorrow.

Can you rebase to HEAD and try again. Not sure why we would see testbench memory failure here as this seem unrelated. The TRACE warning can be ignored.

**Comment by lgirdwood** [comment]
> Mar 25 11:50:31 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware exception
> Mar 25 11:50:31 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: fail in sof_ipc_trace_update_filter -110
> Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx timed out for 0x90050000 (msg/reply size: 8/12)
> Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware exception
> **Mar 25 11:50:34 jhary kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: DMA_TRACE_FREE failed with error: -110**

@jsarha This one is an IPC timeout though, can you check the Zephyr WFI code does the same as SOF
```c
#if (CONFIG_XT_WAITI_DELAY)

static inline void arch_wait_for_interrupt(int level)
{
	int i;

	/* need to make sure the interrupt level won't be lowered */
	if (level)
		panic(SOF_IPC_PANIC_WFI);

	/* this sequence must be atomic on LX6 */
	XTOS_SET_INTLEVEL(5);

	/* LX6 needs a delay */
	for (i = 0; i < 128; i++)
		asm volatile("nop");

	/* and to flush all loads/stores prior to wait */
	asm volatile("isync");
	asm volatile("extw");

	/* now wait */
	asm volatile("waiti 0");
}

#else

static inline void arch_wait_for_interrupt(int level)
{
	/* need to make sure the interrupt level won't be lowered */
	if (level)
		panic(SOF_IPC_PANIC_WFI);

	asm volatile("waiti 0");
}

#endif

static inline void idelay(int n)
{
	while (n--)
		asm volatile("nop");
}

#endif /* __ARCH_LIB_WAIT_H__ */
```

**Comment by lgirdwood** [comment]
@jsarha pls ping @kv2019i when you are back re issue.

**Comment by jsarha** [comment]
Looks like I was tracking a different error. I was tracking this CI-problem that happens also on my up-i11, with or without this PR:
https://sof-ci.01.org/sofpr/PR5582/build12526/devicetest/?model=WHL_UPEXT_HDA_ZEPHYR&testcase=check-suspend-resume-with-capture-5

But then again this problem does not happen on my up-i11:  https://github.com/thesofproject/sof/runs/5664281753?check_suite_focus=true

File: src/arch/xtos-wrapper/include/kernel.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+extern struct tr_ctx wait_tr;
+
+static inline void k_yield(void)
+{
+	tr_dbg(&wait_tr, "WFE");
+#if CONFIG_DEBUG_LOCKS
+	if (lock_dbg_atomic)
+		tr_err_atomic(&wait_tr, "atm");
+#endif
+	platform_wait_for_interrupt(0);
```
Comment (lyakh) [inline src/arch/xtos-wrapper/include/kernel.h:27]: we're trying to recreate the `k_yield()` from Zephyr, right? The description in Zephyr says:
```
/**
 * @brief Yield the current thread.
 *
 * This routine causes the current thread to yield execution to another
 * thread of the same or higher priority. If there are no other ready threads
 * of the same or higher priority, the routine returns immediately.
 */
```
Are you sure WFI is what we need? In Zephyr `k_yield()` will **return immediately** whereas a WFI will suspend the thread until an interrupt. Don't we need something like a call to the scheduler instead?

File: src/arch/xtos-wrapper/include/kernel.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	tr_dbg(&wait_tr, "WFX");
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/arch/xtos-wrapper/include/kernel.h:28]: its a copy exactly from xtos version hence no functional change.

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
@jsarha ping

**Comment by jsarha** [comment]
Apparently all the code using wait_for_interrupt() was xtos specific, so all the other changes  but the first commit, were unnecessary. So only a k_yield() for xtos was added in the end.

**Review by lyakh — COMMENTED** [review]
Do I understand it correctly, that we're now adding an implementation of `k_yield()` to XTOS while there's no current code that has to be converted to it. So, we expect to add code that will be used in XTOS builds and will use `k_yield()`, right? Do we have an idea what kind of code that would be?

**Comment by jsarha** [comment]
> Do I understand it correctly, that we're now adding an implementation of `k_yield()` to XTOS while there's no current code that has to be converted to it. So, we expect to add code that will be used in XTOS builds and will use `k_yield()`, right? Do we have an idea what kind of code that would be?

Short answer: yes, AFAIK

@lgirdwood , can you give a bit longer answer?

File: src/arch/xtos-wrapper/include/kernel.h (hunk 3 above)
Comment (lgirdwood) [reply src/arch/xtos-wrapper/include/kernel.h:27]: Thanks @jsarha it looks like you have found that all wait_for_interrupt() users were xtos based i.e.
```
src/arch/xtensa/schedule/task.c:38:		wait_for_interrupt(0);
src/arch/xtensa/lib/cpu.c:209:		arch_wait_for_interrupt(0);
src/drivers/intel/cavs/sue-ipc.c:58:			wait_for_interrupt(0);
src/drivers/intel/haswell/ipc.c:111:			wait_for_interrupt(0);
src/drivers/amd/renoir/ipc.c:165:			wait_for_interrupt(0);
src/drivers/mediatek/mt8186/ipc.c:89:		wait_for_interrupt(0);
src/drivers/imx/ipc.c:122:			wait_for_interrupt(0);
src/init/init.c:129:		wait_for_interrupt(0);
src/schedule/task.c:57:		wait_for_interrupt(0);
src/platform/intel/cavs/lps_wait.c:101:		arch_wait_for_interrupt(0);
```
Can you confirm that
```
static inline void wait_delay(uint64_t number_of_clks) {}
static inline void wait_delay_ms(uint64_t ms) {}
static inline void wait_delay_us(uint64_t us) {}
```
all have only xtos or driver users (an no audio/pipeline code). If so we can close this and mark it NA.

Comment (jsarha) [reply src/arch/xtos-wrapper/include/kernel.h:27]: There appears to be two instances of wait_delay usage in  src/audio/pipeline/pipeline-stream.c:431 and in src/ipc/ipc4/handler.c:375. The rest of usage is in either under drivers or platform subdirs, e.g. only for implementation XTOS. BTW, the ipc4/handler.c usage has a curious comment on it:
`	while (msg_data.delayed_reply) {
		/* TODO: update to yield(tmeout) for Zephyr */
		wait_delay(10000);

		if (!try_count--) {
			ret = IPC4_FAILURE;
			tr_err(&ipc_tr, "failed to wait schedule thread");
			break;
		}
	}
`
@lgirdwood I did not notice this second usage while we talked,  but reading the comment should we still keep yield() wrapper around to for this (then again yeald() api does not take timeout as a parameter)?

Comment (lgirdwood) [reply src/arch/xtos-wrapper/include/kernel.h:27]: @jsarha looks like https://docs.zephyrproject.org/3.0.0/reference/kernel/threads/index.html#c.k_sleep will fix both.

**Comment by jsarha** [comment]
I'll close this now and will create another one for k_usleep() and wrap the couple wait_delay*() uses behind it.
